packages/toolpad-app/pages/deploy/[appId]/[[...path]].tsx

```diff
@@ -12,6 +12,8 @@ export const getServerSideProps: GetServerSideProps<ToolpadAppProps> = async (co
     import('../../../src/server/basicAuth'),
   ]);
 
+  context.res.removeHeader('X-Frame-Options');
```

Comment:
I assumed that Next.js only starts streaming the response after `getServerSideProps` is called. Considering that you can change the status to be a 404, It seems to be safe.

Reply:
Yes, this should be fine 👍

Reply:
Ah that's a pity, the method of selectively removing the header makes it almost impossible to add optionality to this in the future without completely rewriting it. Another option is to use regex to exclude this header from `/deploy/*` routes:

```rsx
        {
          source: '/((?!deploy/).*)',
          headers: [
            {
              key: 'X-Frame-Options',
              value: 'SAMEORIGIN',
            },
          ],
        },
```

Reply:
@Janpot I didn't understand the "optionality" argument, let me know the approach that you prefer, I will update the PR.

(Personally, I feel that the Next.js headers abstraction doesn't make a lot of sense for non static pages, I might be more intuitive with an express server that would serve the Next.js request handler for some or the pages and set all the headers, request handler per request handler).

Reply:
> I didn't understand the "optionality" argument

Users may not want to allow their apps embedded in iframes. So we could make this header optional. `DENY` by default and provide no header when the user sets the option "Allow embedding in iframes". For that to work, either the header needs to be removed after we fetched the app (meaning we don't see the 404 like you pointed out). Or we need to turn the logic around, where it's not present by default and we enable it conditionally, which is how we built the "public app" feature.

> (Personally, I feel that the Next.js headers abstraction doesn't make a lot of sense for non static pages, I might be more intuitive with an express server that would serve the Next.js request handler for some or the pages, so that we can set all the headers, request handler per request handler).

👍 I agree. With @apedroferreira recently been talking about bringing back the custom server from before. Mainly for different reasons (Next.js edge runtime is just not powerful enough). I don't believe it would solve our problem here. You'd still do something like
```tsx
app.use('*', (req, res, next) => {
  res.setHeader('X-Frame-Options', 'SAMEORIGIN')
  next()
})

app.use('/deploy/*', (req, res, next) => {
  res.removeHeader('X-Frame-Options')
  next()
})

app.use('*', nextHandler)
```
which would have the same effect as setting

```ts
        {
          source: '/((?!deploy/).*)',
          headers: [
            {
              key: 'X-Frame-Options',
              value: 'SAMEORIGIN',
            },
          ],
        },
```
in next config.

Reply:
> Users may not want to allow their apps embedded in iframes. 

Ahh yeah ok. I have missed the part of it's not for everybody to allow iframes. Ok, I'm fixing this with a new commit.

Reply:
I'm going the option that has the safest fail mode: you got the logic wrong, you can't use iframes.

Reply:
Just to be clear. I don't expect this PR to implement it being optional (feel free though 🙂 ). Just that we choose an implementation that has a logical path towards making it so.